File: labs-opensource/M21B-cloud-deployment/starter/verify_connection.js
/**
 * M21B Lab: Connection Verifier (COMPLETE — Node.js)
 * ===================================================
 * Run: node verify_connection.js
 *      OLLAMA_BASE_URL=http://localhost:11434/v1 node verify_connection.js
 */

import OpenAI from "openai";

async function verifyConnection(baseUrl) {
  const url = baseUrl ?? process.env.OLLAMA_BASE_URL ?? "http://localhost:11434/v1";
  const client = new OpenAI({ baseURL: url, apiKey: process.env.API_KEY ?? "ollama" });

  try {
    const response = await client.chat.completions.create({
      model: process.env.MODEL ?? "mistral",
      messages: [{ role: "user", content: "Reply with the single word: connected" }],
      max_tokens: 10,
    });
    const reply = response.choices[0].message.content?.trim();
    console.log(`[OK] Endpoint ${url} responded: ${JSON.stringify(reply)}`);
    console.log(`     Model: ${response.model}`);
    console.log(`     Tokens used: ${response.usage?.total_tokens}`);
  } catch (err) {
    console.error(`[FAIL] Cannot reach endpoint at ${url}: ${err.message}`);
    console.error("       Local:  is Ollama running? (ollama serve)");
    console.error("       Cloud:  is the SSH tunnel up?");
    console.error("       Command: gcloud compute ssh ollama-server -- -L 11434:localhost:11434 -N -f");
    process.exit(1);
  }
}

await verifyConnection();

File: labs-opensource/M21B-cloud-deployment/starter/verify_connection.py
"""
M21B Lab: Connection Verifier (COMPLETE — just run it)
=======================================================
Works identically against local Ollama, an SSH-tunneled cloud VM, or a
managed endpoint — that's the whole point of the OpenAI-compatible pattern.
Run: python verify_connection.py
     OLLAMA_BASE_URL=http://localhost:11434/v1 python verify_connection.py
"""

import os

from openai import OpenAI


def verify_connection(base_url: str | None = None) -> None:
    """Verify Ollama (or any compatible endpoint) is reachable."""
    url = base_url or os.getenv("OLLAMA_BASE_URL", "http://localhost:11434/v1")

    client = OpenAI(base_url=url, api_key=os.getenv("API_KEY", "ollama"))

    try:
        response = client.chat.completions.create(
            model=os.getenv("MODEL", "mistral"),
            messages=[{"role": "user", "content": "Reply with the single word: connected"}],
            max_tokens=10,
        )
        reply = (response.choices[0].message.content or "").strip()
        print(f"[OK] Endpoint {url} responded: {reply!r}")
        print(f"     Model: {response.model}")
        print(f"     Tokens used: {response.usage.total_tokens}")
    except Exception as exc:
        print(f"[FAIL] Cannot reach endpoint at {url}: {exc}")
        print("       Local:  is Ollama running? (ollama serve)")
        print("       Cloud:  is the SSH tunnel up?")
        print("       Command: gcloud compute ssh ollama-server -- -L 11434:localhost:11434 -N -f")
        raise SystemExit(1)


if __name__ == "__main__":
    verify_connection()
